Implement ManagedMediaSource startstreaming/endstreaming event

jyavenard opened the pull request:
#### 77fc3174bf28cd77b65c93acbdf27e622bc2fb3b
<pre>
Implement ManagedMediaSource startstreaming/endstreaming event
<a href="https://bugs.webkit.org/show_bug.cgi?id=253295">https://bugs.webkit.org/show_bug.cgi?id=253295</a>
rdar://106182027

Reviewed by Jer Noble and Youenn Fablet.

We set a low and high watermarks of respectively 10 and 30s ; we continuously
monitor currentTime progress to ensure that buffering is within those
two thresholds.
Upcoming change would make those configurable.

* LayoutTests/media/media-source/media-managedmse-streaming-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-streaming.html: Added.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::startStreaming):
(WebCore::ManagedMediaSource::endStreaming):
(WebCore::ManagedMediaSource::isBuffered const):
(WebCore::ManagedMediaSource::monitorSourceBuffers):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::updateBufferedIfNeeded):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::removeTimerFired):
(WebCore::SourceBuffer::memoryPressure):
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::monitorSourceBuffers):
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:

Canonical link: <a href="https://commits.webkit.org/261232@main">https://commits.webkit.org/261232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6edd2fac8f27325731d049f2be9ea2238fec5043

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110917 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20056 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119767 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114877 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21441 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2026 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116669 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15929 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99103 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103361 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97889 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30768 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44308 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12576 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32104 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13128 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9086 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18534 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7794 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15073 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
```diff
@@ -63,5 +70,85 @@ bool ManagedMediaSource::isTypeSupported(ScriptExecutionContext& context, const
+    if (m_streaming)
+        return;
+    m_streaming = true;
+    scheduleEvent(eventNames().startstreamingEvent);
```
There is a small risk of quickly enqueuing 3 events, say startstreaming/endstreaming/startstreaming.
I guess the threshold should be high enough it does not happen in practice and we can always revisit this if needed.

jyavenard replied:
If the player deleted buffered data as soon as `endstreaming` was fired, I don't see how we could avoid this.
And is it a problem?

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
```diff
@@ -28,12 +28,19 @@
 
 #if ENABLE(MANAGED_MEDIA_SOURCE) && ENABLE(MEDIA_SOURCE)
 
+#include "Event.h"
```
Do we need this one?

LGunarM replied:
No

jyavenard committed:
Implement ManagedMediaSource startstreaming/endstreaming event
https://bugs.webkit.org/show_bug.cgi?id=253295
rdar://106182027

Reviewed by Jer Noble and Youenn Fablet.

We set a low and high watermarks of respectively 10 and 30s ; we continuously
monitor currentTime progress to ensure that buffering is within those
two thresholds.
Upcoming change would make those configurable.

* LayoutTests/media/media-source/media-managedmse-streaming-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-streaming.html: Added.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::startStreaming):
(WebCore::ManagedMediaSource::endStreaming):
(WebCore::ManagedMediaSource::isBuffered const):
(WebCore::ManagedMediaSource::monitorSourceBuffers):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::updateBufferedIfNeeded):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::removeTimerFired):
(WebCore::SourceBuffer::memoryPressure):
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::monitorSourceBuffers):
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:

Canonical link: https://commits.webkit.org/261232@main
